Implement streaming for Bevy Remote

notmd opened the pull request:
# Objective

- Implement streaming for Bevy Remote via WebSocket to allow the client subscribing to listen for change without having to continuously poll the server.

## Solution

- Jsonrpc spec doesn't have streaming so I implement a similar solution to it. 
Due to WebSocket doesn't have a request body, every request **MUST** contain a `body` query encoded in JSON of `BrpRequest`. This is similar to [trpc](https://trpc.io/docs/websockets). So the URI should be something like this
```
ws://localhost:15702?body={"method": "example/stream","jsonrpc":"2.0","params":{"some_thing":"value"}}
```
For now, the streaming is a one-way connection from server to client (similar to SSE).

## Testing

- I update the example to demonstrate this.  Run with the --stream flag to test it.
```
cargo run --example client -- --stream
```
---


## Question
- This feature pulls quit a lot of dependencies. Should we hide it behind a feature flag?

## Migration Guide

notmd commented on crates/bevy_remote/Cargo.toml:
```diff
@@ -27,6 +27,10 @@ hyper = { version = "1", features = ["server", "http1"] }
 http-body-util = "0.1"
+tungstenite = "0.24"
+hyper-tungstenite = "0.15"
+futures-util = "0.3"
```
I cannot find any existing bevy dependency that has the [send](https://docs.rs/futures-util/latest/futures_util/sink/trait.SinkExt.html#method.send) method so I have to pull `futures-util` in.

JMS55 commented:
Curious to hear your thoughts on websockets over webtransport, quic, or just UDP. I'm not too plugged into the BRP design, but looking at the document it has it's own ordering mechanism, making reliable transport unnecessary.

notmd commented:
I choose WebSocket due to its simplicity and popularity, but yeah we can add other backends in the future.

BenjaminBrienen commented on crates/bevy_remote/src/lib.rs:
```diff
@@ -866,3 +961,71 @@ async fn process_single_request(
+        .query()
+        .map(|query| {
+            // Simple query string parsing
+            let mut map = HashMap::new();
```
Can we refactor this query string parsing logic to be in its own module so that it can be used in other places as well? It seems like a generically useful piece of logic.

notmd commented:
Marking this as draft now, the disconnect part is not properly handled.

notmd commented:
Sorry for the large diff in the last commit. In order to properly clean up disconnected clients I have moved streaming messages to a separate channel and introduced `BrpClientId` to track between 2 threads, also all the logic to a separate module. The `BrpClientId` is also useful for users to handle cleanup as well. I will expose it in the follow-up PR.

notmd commented on crates/bevy_remote/src/lib.rs:
```diff
@@ -737,54 +777,68 @@ async fn server_main(
+    let listener = Async::<TcpListener>::bind((address, port))?;
+    let mut client_id: usize = 0;
     loop {
-        let (client, _) = listener.accept().await?;
-
+        let (stream, _) = listener.accept().await?;
+        client_id = client_id.wrapping_add(1);
```
This simple id generator should be ok for a debug tool right?

notmd commented:
With the http logic is now in a separate plugin in #15438, I think the streaming logic can be the same, and still reuse the builtins method very much. So I will close this and experiment with streaming first and decide whether to upstream to Bevy.
